File: src/main/java/org/arbor/gtnn/api/machine/multiblock/LargeNaquadahReactorMachine.java
package org.arbor.gtnn.api.machine.multiblock;

import com.gregtechceu.gtceu.api.capability.recipe.EURecipeCapability;
import com.gregtechceu.gtceu.api.capability.recipe.IO;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.feature.IExplosionMachine;
import com.gregtechceu.gtceu.api.machine.feature.multiblock.IMultiPart;
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableElectricMultiblockMachine;
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableMultiblockMachine;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.RecipeHelper;
import com.gregtechceu.gtceu.api.recipe.content.Content;
import com.gregtechceu.gtceu.common.data.GTMaterials;
import com.gregtechceu.gtceu.common.machine.multiblock.part.FluidHatchPartMachine;
import com.lowdragmc.lowdraglib.side.fluid.FluidStack;
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted;
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder;
import it.unimi.dsi.fastutil.longs.Long2ObjectMaps;
import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.world.level.material.Fluid;
import org.arbor.gtnn.data.GTNNMaterials;
import org.arbor.gtnn.data.GTNNRecipeTypes;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class LargeNaquadahReactorMachine extends WorkableElectricMultiblockMachine implements IExplosionMachine {
    private static final ManagedFieldHolder MANAGED_FIELD_HOLDER = new ManagedFieldHolder(LargeNaquadahReactorMachine.class, WorkableMultiblockMachine.MANAGED_FIELD_HOLDER);
    private static final Map<Fluid, Integer> activeFluid = Map.of(
            GTMaterials.Caesium.getFluid(), 2,
            GTMaterials.Uranium235.getFluid(), 3,
            GTMaterials.Naquadah.getFluid(), 4
    );
    private static final Map<Fluid, Integer> activeFluidCost = Map.of(
            GTMaterials.Caesium.getFluid(), 180,
            GTMaterials.Uranium235.getFluid(), 180,
            GTMaterials.Naquadah.getFluid(), 20
    );
    private final List<Fluid> fuelFluids = List.of(
            GTNNMaterials.ThoriumBasedLiquidFuelExcited.getFluid(),
            GTNNMaterials.UraniumBasedLiquidFuelExcited.getFluid(),
            GTNNMaterials.PlutoniumBasedLiquidFuelExcited.getFluid()
    );
    private Set<FluidHatchPartMachine> hatchPartMachines;
    @Persisted
    private boolean hasAir = false;
    @Persisted
    private boolean hasCool = false;
    @Persisted
    private int activeFluidPower = 1;
    private Fluid lockFluid = null;

    public LargeNaquadahReactorMachine(IMachineBlockEntity holder) {
        super(holder);
    }

    @Nullable
    public static GTRecipe modifyRecipe(MetaMachine machine, GTRecipe recipe) {
        if (!(recipe.recipeType == GTNNRecipeTypes.LARGE_NAQUADAH_REACTOR_RECIPES)) return null;
        if (machine instanceof LargeNaquadahReactorMachine largeNaquadahReactorMachine) {
            final int duration = recipe.duration;
            checkHatch(largeNaquadahReactorMachine, duration);
            GTRecipe copyRecipe = recipe.copy();
            if (!largeNaquadahReactorMachine.hasAir) {
                copyRecipe.tickOutputs.clear();
                copyRecipe.outputs.clear();
                return copyRecipe;
            }
            long eut = RecipeHelper.getOutputEUt(copyRecipe);
            if (largeNaquadahReactorMachine.hasCool) {
                eut = (long) (eut * 1.5);
            }
            eut *= largeNaquadahReactorMachine.activeFluidPower;
            copyRecipe.tickOutputs.put(EURecipeCapability.CAP, List.of(new Content(eut, 1.0f, 0.0f, null, null)));
            return copyRecipe;
        }
        return null;
    }

    private static void checkHatch(LargeNaquadahReactorMachine largeNaquadahReactorMachine, int duration) {
        largeNaquadahReactorMachine.hasCool = false;
        largeNaquadahReactorMachine.hasAir = false;
        largeNaquadahReactorMachine.activeFluidPower = 1;
        for (var hatch : largeNaquadahReactorMachine.hatchPartMachines) {
            var tank = hatch.tank;
            var io = tank.getHandlerIO();
            if (io == IO.IN || io == IO.BOTH) {
                for (int i = 0; i < tank.getTanks(); i++) {
                    final FluidStack fluid = tank.storages[i].getFluid();
                    checkLockFluid(largeNaquadahReactorMachine, fluid);
                    active(largeNaquadahReactorMachine, fluid, duration);
                    if (cool(fluid, duration)) largeNaquadahReactorMachine.hasCool = true;
                    if (air(fluid, duration)) largeNaquadahReactorMachine.hasAir = true;
                }
            }
        }
    }

    private static void checkLockFluid(LargeNaquadahReactorMachine machine, FluidStack fluid) {
        if (machine.fuelFluids.contains(fluid.getFluid())) {
            if (machine.lockFluid == null) {
                machine.lockFluid = fluid.getFluid();
            } else if (machine.lockFluid != fluid.getFluid()) {
                machine.doExplosion(4 * 32);
            }
        }
    }


    //////////////////////////////////////
    //******     RECIPE LOGIC    *******//
    //////////////////////////////////////

    private static boolean air(FluidStack fluid, int duration) {
        if (fluid.getFluid().isSame(GTMaterials.LiquidAir.getFluid())) {
            final int airAmount = 2400 / 20 * duration;
            if (fluid.getAmount() >= airAmount) {
                fluid.setAmount(fluid.getAmount() - airAmount);
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    private static void active(LargeNaquadahReactorMachine machine, FluidStack fluid, int duration) {
        if (activeFluid.containsKey(fluid.getFluid())) {
            final int activeFluidCostI = activeFluidCost.get(fluid.getFluid()) / 20 * duration;
            final int activeFluidPower = activeFluid.get(fluid.getFluid());
            if (machine.activeFluidPower <= activeFluidPower && fluid.getAmount() >= activeFluidCostI) {
                machine.activeFluidPower = activeFluidPower;
                fluid.setAmount(fluid.getAmount() - activeFluidCostI);
            }
        }
    }

    private static boolean cool(FluidStack fluid, int duration) {
        if (fluid.getFluid().isSame(GTMaterials.PCBCoolant.getFluid())) {
            final int coldAmount = 1000 / 20 * duration;
            if (fluid.getAmount() >= coldAmount) {
                fluid.setAmount(fluid.getAmount() - coldAmount);
                return true;
            }
        }
        return false;
    }

    //////////////////////////////////////
    //***    Multiblock LifeCycle    ***//
    //////////////////////////////////////
    @Override
    public void onStructureFormed() {
        super.onStructureFormed();
        var matchContext = getMultiblockState().getMatchContext();
        Map<Long, IO> ioMap = matchContext.getOrCreate("ioMap", Long2ObjectMaps::emptyMap);
        // Cache the result of getParts() to prevent repetitive calls
        List<IMultiPart> parts = getParts();
        for (IMultiPart part : parts) {
            IO io = ioMap.getOrDefault(part.self().getPos().asLong(), IO.BOTH);
            if (io == IO.NONE) continue;
            if (part instanceof FluidHatchPartMachine fluidHatchPartMachine) {
                hatchPartMachines = APartAbility.getOrDefault(hatchPartMachines, HashSet::new);
                hatchPartMachines.add(fluidHatchPartMachine);
            }
        }
    }

    @Override
    public void onStructureInvalid() {
        super.onStructureInvalid();
        hatchPartMachines = null;
        lockFluid = null;
    }

    @Override
    public boolean alwaysTryModifyRecipe() {
        return true;
    }

    public double getFinalPowerRate() {
        double activeFluidPower = this.activeFluidPower;
        if (hasCool) activeFluidPower *= 1.5;
        return activeFluidPower;
    }

    //////////////////////////////////////
    //******       NBT SAVE      *******//
    //////////////////////////////////////
    @Override
    public ManagedFieldHolder getFieldHolder() {
        return MANAGED_FIELD_HOLDER;
    }
}

File: src/main/kotlin/dev/arbor/gtnn/common/machine/multiblock/LargeNaquadahReactorMachine.kt
package dev.arbor.gtnn.common.machine.multiblock

import com.gregtechceu.gtceu.api.capability.recipe.EURecipeCapability
import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.MetaMachine
import com.gregtechceu.gtceu.api.machine.feature.IExplosionMachine
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableElectricMultiblockMachine
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableMultiblockMachine
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.content.Content
import com.gregtechceu.gtceu.api.recipe.ingredient.EnergyStack
import com.gregtechceu.gtceu.api.recipe.modifier.ModifierFunction
import com.gregtechceu.gtceu.common.data.GTMaterials
import com.gregtechceu.gtceu.common.machine.multiblock.part.FluidHatchPartMachine
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import dev.arbor.gtnn.data.GTNNMaterials
import dev.arbor.gtnn.data.GTNNRecipeTypes
import dev.arbor.gtnn.extension.NNUtils
import it.unimi.dsi.fastutil.longs.Long2ObjectMap
import it.unimi.dsi.fastutil.longs.Long2ObjectMaps
import net.minecraft.MethodsReturnNonnullByDefault
import net.minecraft.world.level.material.Fluid
import net.minecraftforge.fluids.FluidStack
import javax.annotation.ParametersAreNonnullByDefault

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
class LargeNaquadahReactorMachine(holder: IMachineBlockEntity) : WorkableElectricMultiblockMachine(holder),
    IExplosionMachine {
    private val activeFluid: Map<Fluid, Int> = mapOf(
        GTMaterials.Caesium.fluid to 2,
        GTMaterials.Uranium235.fluid to 3,
        GTMaterials.Naquadah.fluid to 4
    )
    private val activeFluidCost: Map<Fluid, Int> = mapOf(
        GTMaterials.Caesium.fluid to 180,
        GTMaterials.Uranium235.fluid to 180,
        GTMaterials.Naquadah.fluid to 20
    )
    private val fuelFluids: List<Fluid> = listOf(
        GTNNMaterials.ThoriumBasedLiquidFuelExcited.fluid,
        GTNNMaterials.UraniumBasedLiquidFuelExcited.fluid,
        GTNNMaterials.PlutoniumBasedLiquidFuelExcited.fluid
    )
    private var hatchPartMachines: Set<FluidHatchPartMachine>? = null

    @Persisted
    private var hasAir = false

    @Persisted
    private var hasCool = false

    @Persisted
    private var activeFluidPower = 1
    private var lockFluid: Fluid? = null

    private fun checkHatch(machine: LargeNaquadahReactorMachine, duration: Int) {
        machine.hasCool = false
        machine.hasAir = false
        machine.activeFluidPower = 1
        for (hatch in machine.hatchPartMachines!!) {
            val tank = hatch.tank
            val io = tank.getHandlerIO()
            if (io == IO.IN || io == IO.BOTH) {
                for (i in tank.storages) {
                    val fluid = i.fluid
                    if (i.fluid == FluidStack.EMPTY) continue
                    checkLockFluid(machine, fluid)
                    active(machine, fluid, duration)
                    if (cool(fluid, duration)) machine.hasCool = true
                    if (air(fluid, duration)) machine.hasAir = true
                }
            }
            if (machine.hasCool && machine.hasAir && machine.activeFluidPower != 1) {
                return
            }
        }
    }

    private fun checkLockFluid(machine: LargeNaquadahReactorMachine, fluid: FluidStack) {
        if (machine.fuelFluids.contains(fluid.fluid)) {
            if (machine.lockFluid == null) {
                machine.lockFluid = fluid.fluid
            } else if (machine.lockFluid != fluid.fluid) {
                machine.doExplosion((4 * 32).toFloat())
            }
        }
    }


    //////////////////////////////////////
    //******     RECIPE LOGIC    *******//
    //////////////////////////////////////

    private fun air(fluid: FluidStack, duration: Int): Boolean {
        if (fluid.fluid.isSame(GTMaterials.LiquidAir.fluid)) {
            val airAmount = 2400 / 20 * duration
            if (fluid.amount >= airAmount) {
                fluid.amount -= airAmount
                return true
            } else {
                return false
            }
        } else {
            return false
        }
    }

    private fun active(machine: LargeNaquadahReactorMachine, fluid: FluidStack, duration: Int) {
        if (activeFluid.containsKey(fluid.fluid)) {
            val activeFluidCostI = activeFluidCost[fluid.fluid]!! / 20 * duration
            val activeFluidPower = activeFluid[fluid.fluid]
            if (machine.activeFluidPower <= activeFluidPower!! && fluid.amount >= activeFluidCostI) {
                machine.activeFluidPower = activeFluidPower
                fluid.amount -= activeFluidCostI
            }
        }
    }

    private fun cool(fluid: FluidStack, duration: Int): Boolean {
        if (fluid.fluid.isSame(GTMaterials.PCBCoolant.fluid)) {
            val coldAmount = 1000 / 20 * duration
            if (fluid.amount >= coldAmount) {
                fluid.amount -= coldAmount
                return true
            }
        }
        return false
    }

    //////////////////////////////////////
    //***    Multiblock LifeCycle    ***//
    //////////////////////////////////////
    override fun onStructureFormed() {
        super.onStructureFormed()
        val matchContext = multiblockState.matchContext
        val ioMap = matchContext.getOrCreate<Long2ObjectMap<Any>>("ioMap", Long2ObjectMaps::emptyMap)
        // Cache the result of getParts() to prevent repetitive calls
        val parts = getParts()
        for (part in parts) {
            val io = ioMap.getOrDefault(part.self().pos.asLong(), IO.BOTH)
            if (io == IO.NONE) continue
            if (part is FluidHatchPartMachine) {
                hatchPartMachines = NNUtils.getOrDefault(hatchPartMachines, ::mutableSetOf)
                (hatchPartMachines as MutableSet).add(part)
            }
        }
    }

    override fun onStructureInvalid() {
        super.onStructureInvalid()
        hatchPartMachines = hashSetOf()
        lockFluid = null
    }

    override fun alwaysTryModifyRecipe(): Boolean {
        return true
    }

    fun getFinalPowerRate(): Int {
        var activeFluidPower = this.activeFluidPower
        if (hasCool) {
            activeFluidPower = (activeFluidPower * 1.5).toInt()
        }
        return activeFluidPower
    }

    //////////////////////////////////////
    //******       NBT SAVE      *******//
    //////////////////////////////////////
    override fun getFieldHolder(): ManagedFieldHolder {
        return MANAGED_FIELD_HOLDER
    }

    companion object {
        private val MANAGED_FIELD_HOLDER =
            ManagedFieldHolder(LargeNaquadahReactorMachine::class.java, WorkableMultiblockMachine.MANAGED_FIELD_HOLDER)

        fun modifyRecipe(machine: MetaMachine, recipe: GTRecipe): ModifierFunction {
            if (recipe.recipeType != GTNNRecipeTypes.LARGE_NAQUADAH_REACTOR_RECIPES) return ModifierFunction.NULL
            if (machine is LargeNaquadahReactorMachine) {
                return ModifierFunction {
                    val duration = recipe.duration
                    machine.checkHatch(machine, duration)
                    val copyRecipe = recipe.copy()
                    if (!machine.hasAir) {
                        copyRecipe.tickOutputs.clear()
                        copyRecipe.outputs.clear()
                        return@ModifierFunction copyRecipe
                    }
                    var eu = copyRecipe.outputEUt.totalEU
                    if (machine.hasCool) {
                        eu = (eu * 1.5).toLong()
                    }
                    eu *= machine.activeFluidPower
                    val eut = EnergyStack(eu)
                    copyRecipe.tickOutputs[EURecipeCapability.CAP] = listOf(Content(eut, 1, 1, 0))
                    return@ModifierFunction copyRecipe
                }
            }
            return ModifierFunction.NULL
        }
    }
}
